### PR TITLE
0066-VALW-validator_tendermint_weights - post review new test coverage

### PR DESCRIPTION
**Pull request by claudiumilea**

### CODE REVIEW

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 1]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
 ### Basic sanity check (<a name="0066-VALW-001" href="#0066-VALW-001">0066-VALW-001</a>)
 1. set up a network with 5 validators 
-1. give each of the validators the following number of self-staked tokens: `2000`
-1. wait for the delegation to become active in the next epoch
-1. when the epoch begins verify that the delegation went through
-1. check that Tendermint weights for the validators are `2000`
+2. give each of the validators the following number of self-staked tokens: `2000`
+3. wait for the delegation to become active in the next epoch
+4. when the epoch begins verify that the delegation went through
+5. check that Tendermint weights for the validators are `2000`
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]: I don't see what the change was here?

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 2]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
+
 
 ### Non-uniform stake check (<a name="0066-VALW-002" href="#0066-VALW-002">0066-VALW-002</a>)
 1. set up a network with 5 validators
-1. give the first `4` validators `1000` of self-stake each. Give the last validator `500`. 
-1. wait for the delegation to become active in the next epoch
-1. check that Tendermint weights for the first `4` validators are `2222` and the last validator `1111`. 
+2. give the first `4` validators `1000` of self-stake each. Give the last validator `500`. 
+3. wait for the delegation to become active in the next epoch
+4. check that Tendermint weights for the first `4` validators are `2222` and the last validator `1111`.
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]: And here?

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 3]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
+5. now every validator removes (via undelegate now) their stake and there are no delegations from other parties
+6. wait for 1000 blocks to pass
+7. the Tendermint weights for each validator are `10` each as there is no stake in the network. 
+
+### Validator has 1000 stake but 500 delegated , check only attention to 500 (<a name="0066-VALW-006" href="#0066-VALW-006">0066-VALW-006</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
+3. delegate for the first `4` validators `1000` and only delegate `200` to the last node
+4. wait for the delegation to become active in the next epoch
+5. check that Tendermint weights for the first `4` validators are `2222` and the last validator `11119`.
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]: These numbers don't make sense to me. Why would the last validator have ~ 5x as much as the first four? Also the Tendermint weights will depend on performance (so say we're assuming this is 1 for all) and on all parameters that impact `validatorScore` in https://github.com/vegaprotocol/specs/blob/master/protocol/0061-REWP-pos_rewards.md .

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 4]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
+
+### Validators without self-delegated, check  (<a name="0066-VALW-007" href="#0066-VALW-007">0066-VALW-007</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `0`
+3. wait for the delegation to become active in the next epoch
+4. check the Tendermint weights
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]: This doesn't seem to be doing anything.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 5]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
+### Validator has 1000 stake but 500 delegated , check only attention to 500 (<a name="0066-VALW-006" href="#0066-VALW-006">0066-VALW-006</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
+3. delegate for the first `4` validators `1000` and only delegate `200` to the last node
+4. wait for the delegation to become active in the next epoch
+5. check that Tendermint weights for the first `4` validators are `2222` and the last validator `11119`.
+
+### Validators without self-delegated, check  (<a name="0066-VALW-007" href="#0066-VALW-007">0066-VALW-007</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `0`
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:105]: What's the network parameter setting self-delegation minimum?

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 6]
```diff
@@ -15,38 +15,69 @@ If the network has no stake at all, then all validators would have equal voting
...
+3. wait for the delegation to become active in the next epoch
+4. check the Tendermint weights
+
+### Validators delegating to other validators (<a name="0066-VALW-008" href="#0066-VALW-008">0066-VALW-008</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
+3. delegate for the first `4` validators `1000` and from the 5th validator delegate `1000` to the 1st validator
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:119]: What does it mean for the 5th validator to delegate 1000 to the 1st validator? Meaning that they'll have `0` own stake? And what should the outcome be on "check the Tendermint weights" below?

**Review by davidsiska-vega — CHANGES_REQUESTED** [review]

**Comment by vega-paul** [comment]
WE are unsure what the scores will be @davidsiska-vega so will produce the numbers and check back with you on what we find.

**Comment by vega-paul** [comment]
A few mods to the new ACs. Tests are being written, currently we just dont know what the scores will be from the new scenarios so will report back when we have output @davidsiska-vega

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 7]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
 
 ### Basic sanity check (<a name="0066-VALW-001" href="#0066-VALW-001">0066-VALW-001</a>)
 1. set up a network with 5 validators 
-1. give each of the validators the following number of self-staked tokens: `2000`
-1. wait for the delegation to become active in the next epoch
-1. when the epoch begins verify that the delegation went through
-1. check that Tendermint weights for the validators are `2000`
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:21]: I don't follow why we're removing the weights. The point is that 5 validators with the same amount of tokens staked and delegated should have the same weight. The weight is an integer (out of 10 000).

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 8]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+2. give each of the validators the following number of self-staked tokens: `2000`
+3. wait for the delegation to become active in the next epoch
+4. when the epoch begins verify that the delegation went through
+
 
 ### Non-uniform stake check (<a name="0066-VALW-002" href="#0066-VALW-002">0066-VALW-002</a>)
 1. set up a network with 5 validators
-1. give the first `4` validators `1000` of self-stake each. Give the last validator `500`. 
-1. wait for the delegation to become active in the next epoch
-1. check that Tendermint weights for the first `4` validators are `2222` and the last validator `1111`.
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:27]: Ditto.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 9]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
-1. check that Tendermint weights for the first `4` validators are `2222` and the last validator `1111`. 
+2. give the first `4` validators `1000` of self-stake each. Give the last validator `500`. 
+3. wait for the delegation to become active in the next epoch
+
 
 ### Zero stake check  (<a name="0066-VALW-003" href="#0066-VALW-003">0066-VALW-003</a>)
 1. set up a network with 5 validators
-1. give the first `4` validators `1000` of self-stake each. Give the last validator `0`. 
-1. wait for the delegation to become active in the next epoch
-1. check that Tendermint weights for the first `4` validators are `2500` and the last validator `1`.
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:33]: Ditto.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 10]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
-1. check that Tendermint weights for the first `4` validators are `2500` and the last validator `1`. 
-1. just before the epoch ends the last validator self-stakes `500`. 
-1. next epoch starts 
-1. we see that Tendermint weights for the first `4` validators are `2222` and the last validator `1111`. 
+2. give the first `4` validators `1000` of self-stake each. Give the last validator `0`. 
+3. wait for the delegation to become active in the next epoch
+5. check that Tendermint weights
+6. just before epoch 0 ends the last validator self-stakes `500`. 
+7. epoch 1 starts 
+5. check that Tendermint weights
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]: What should be the outcome of the "check"? We can always "check" the weights - they will be an integer. The test should verify they're the right integer.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 11]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
-1. check that Tendermint weights for the validators are `2000`
-1. now every validator removes (via undelegate now) their stake and there are no delegations from other parties
-1. wait for the epoch to end
-1. the Tendermint weights for each validator are `10` each as there is no stake in the network. 
+2. give each of the validators the following number of self-staked tokens: `2000`
+3. wait for the delegation to become active in the next epoch
+4. check that Tendermint weights for the validators are `2000`
+5. now every validator removes (via undelegate now) their stake and there are no delegations from other parties
+6. wait for 1000 blocks to pass
+7. the Tendermint weights for each validator are `10` each as there is no stake in the network.
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:87]: Why is this changed? I don't see anything different.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 12]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+3. delegate to all validators `500` tokens`
+4. wait for the delegation to become active in the next epoch
+5. check that Tendermint weights 
+
+### Validators without self-delegated, check  (<a name="0066-VALW-007" href="#0066-VALW-007">0066-VALW-007</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `0`
+3. For several parties create stake and delegate to validators
+4. wait for the delegation to become active in the next epoch
+5. check the Tendermint weights
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:65]: To see what?

**Review by davidsiska-vega — CHANGES_REQUESTED** [review]

**Comment by davidsiska-vega** [comment]
> WE are unsure what the scores will be @davidsiska-vega so will produce the numbers and check back with you on what we find.

Who is we? I would imagine e.g. @ze97286 will tell you what the weights would be and my understanding is that they should be what is in the spec before the PR.

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 12 above)
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md]:
```suggestion
5. check the Tendermint weights to see every validator should have weight `1`.  
```

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 13]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+### Validators without self-delegated, check  (<a name="0066-VALW-007" href="#0066-VALW-007">0066-VALW-007</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `0`
+3. For several parties create stake and delegate to validators
+4. wait for the delegation to become active in the next epoch
+5. check the Tendermint weights 
+
+### Validators delegating to other validators (<a name="0066-VALW-008" href="#0066-VALW-008">0066-VALW-008</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:118]: what is the minimum self stake parameter?

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 14]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+3. delegate for the first `4` validators `1000` and from the 5th validator delegate `1000` to the 1st validator
+4. wait for the delegations to become active in the next epoch
+5. check the Tendermint weights across all validators
+
+### Validator delegate and undelegate in the same epoch (<a name="0066-VALW-009" href="#0066-VALW-009">0066-VALW-009</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:132]: what is the minimum self stake?

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 15]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+3. delegate in the same epoch
+4. undelegate all in the same epoch
+6. wait for the next epoch
+7. Check that Tendermint weights whereby validators would have equal voting power of 10
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:136]:
```suggestion
7. Check that Tendermint weights whereby validators would have equal voting power of 2000
```

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 5 above)
Comment (vega-paul) [reply protocol/0066-VALW-validator_tendermint_weights.md:105]: If you mean  "reward.staking.delegation.minimumValidatorStake": "0", 
then its default is 0 and remains 0 for the tests.

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 6 above)
Comment (vega-paul) [reply protocol/0066-VALW-validator_tendermint_weights.md:119]: yes thats correct, 5th val delegates to val 1

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 14 above)
Comment (vega-paul) [reply protocol/0066-VALW-validator_tendermint_weights.md:132]: 0

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 13 above)
Comment (vega-paul) [reply protocol/0066-VALW-validator_tendermint_weights.md:118]: 0

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 7 above)
Comment (vega-paul) [reply protocol/0066-VALW-validator_tendermint_weights.md:21]: If you would like to add the expected weights I can verify against the results we see. Currently all results are calculated based on rankingScore values and values from tendermint and these are used to verify results.

Comment (davidsiska-vega) [reply protocol/0066-VALW-validator_tendermint_weights.md:21]: > Currently all results are calculated based on rankingScore values and values from tendermint and these are used to verify results.

Then let's add the formula here and say that the formula give the right answer.

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 5 above)
Comment (davidsiska-vega) [reply protocol/0066-VALW-validator_tendermint_weights.md:105]: Can we make that clear by mentioning it in the AC? Someone reading the spec AC shouldn't need to know the default value of any parameter.

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 13 above)
Comment (davidsiska-vega) [reply protocol/0066-VALW-validator_tendermint_weights.md:118]: As above.

File: protocol/0066-VALW-validator_tendermint_weights.md [hunk 16]
```diff
@@ -15,37 +15,67 @@ If the network has no stake at all, then all validators would have equal voting
...
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `0`
+3. For several parties create stake and delegate to validators
+4. wait for the delegation to become active in the next epoch
+5. check the Tendermint weights 
+
+### Validators delegating to other validators (<a name="0066-VALW-008" href="#0066-VALW-008">0066-VALW-008</a>)
+1. set up a network with 5 validators
+2. give each of the validators the following number of self-staked tokens: `1000`
+3. delegate for the first `4` validators `1000` and from the 5th validator delegate `1000` to the 1st validator
```
Comment (davidsiska-vega) [inline protocol/0066-VALW-validator_tendermint_weights.md:119]:
```suggestion
3. self-delegate for the first `4` validators `1000` and from the vega public key associated to the 5th validator delegate `1000` to the 1st validator. 
```

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 14 above)
Comment (davidsiska-vega) [reply protocol/0066-VALW-validator_tendermint_weights.md:132]: As above.

File: protocol/0066-VALW-validator_tendermint_weights.md (hunk 7 above)
Comment (gordsport) [reply protocol/0066-VALW-validator_tendermint_weights.md:21]: @vega-paul - please can you update the AC and add the formulas and numbers. We can then test this, review the numbers and decide the next steps?

**Comment by vega-paul** [comment]
@davidsiska-vega All results from the tests have been added to this spec, be aware that the scores included in this doc match the expected result and so any if any results here look incorrect we would need to revisit the test and identify a possible core/test bug.

**Review by davidsiska-vega — APPROVED** [review]
Amazing, thank you!!!
